model/modelpb/apmevent.pb.json.go

```diff
@@ -0,0 +1,279 @@
+// Licensed to Elasticsearch B.V. under one or more contributor
+// license agreements. See the NOTICE file distributed with
+// this work for additional information regarding copyright
+// ownership. Elasticsearch B.V. licenses this file to you under
+// the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package modelpb
+
+import (
+	"github.com/elastic/apm-data/model/internal/modeljson"
+	"go.elastic.co/fastjson"
+)
+
+func (e *APMEvent) MarshalJSON() ([]byte, error) {
+	var w fastjson.Writer
+	if err := e.MarshalFastJSON(&w); err != nil {
+		return nil, err
+	}
+	return w.Bytes(), nil
+}
+
+func (e *APMEvent) MarshalFastJSON(w *fastjson.Writer) error {
+	out := modeljson.Document{
+		Span: &modeljson.Span{
+			Message:     &modeljson.Message{},
+			Composite:   &modeljson.SpanComposite{},
+			Destination: &modeljson.SpanDestination{},
+			DB:          &modeljson.DB{},
+		},
+		Transaction: &modeljson.Transaction{
+			UserExperience: &modeljson.UserExperience{},
+			Message:        &modeljson.Message{},
+		},
+		Metricset: &modeljson.Metricset{},
+		Error: &modeljson.Error{
+			Exception: &modeljson.Exception{},
+			Log:       &modeljson.ErrorLog{},
+		},
+		TimestampStruct: &modeljson.Timestamp{},
+		Cloud:           &modeljson.Cloud{},
+		Service: &modeljson.Service{
+			Node:      &modeljson.ServiceNode{},
+			Language:  &modeljson.Language{},
+			Runtime:   &modeljson.Runtime{},
+			Framework: &modeljson.Framework{},
+			Origin:    &modeljson.ServiceOrigin{},
+			Target:    &modeljson.ServiceTarget{},
+		},
+		FAAS:       &modeljson.FAAS{},
+		Network:    &modeljson.Network{},
+		Container:  &modeljson.Container{},
+		User:       &modeljson.User{},
+		Device:     &modeljson.Device{},
+		Kubernetes: &modeljson.Kubernetes{},
+		Observer:   &modeljson.Observer{},
+		Agent:      &modeljson.Agent{},
+		HTTP: &modeljson.HTTP{
+			Request: &modeljson.HTTPRequest{
+				Body: &modeljson.HTTPRequestBody{},
+			},
+			Response: &modeljson.HTTPResponse{},
+		},
+		UserAgent: &modeljson.UserAgent{},
+		Parent:    &modeljson.Parent{},
+		Trace:     &modeljson.Trace{},
+		Host: &modeljson.Host{
+			OS: &modeljson.OS{},
+		},
+		URL:         &modeljson.URL{},
+		Log:         &modeljson.Log{},
+		Source:      &modeljson.Source{},
+		Client:      &modeljson.Client{},
+		Child:       &modeljson.Child{},
+		Destination: &modeljson.Destination{},
+		Session:     &modeljson.Session{},
+		Process:     &modeljson.Process{},
+		Event:       &modeljson.Event{},
```

Comment:
Won't this cause a lot of empty struct fields in the JSON output? It will also lead to unnecessary allocations.

Reply:
> Won't this cause a lot of empty struct fields in the JSON output? 

This should be fixed, mapping will take care of setting them to nil.

> It will also lead to unnecessary allocations.

No, allocations decreased because these are all allocated on the stack. Before they were escaping to the heap so we had more allocations. See the benchmark above.

Reply:
> This should be fixed, mapping will take care of setting them to nil.

:+1: I missed that we were setting them to nil.

> No, allocations decreased because these are all allocated on the stack. Before they were escaping to the heap so we had more allocations. See the benchmark above.

Ah, I think the issue is the nested objects. The outer types declared should be on the stack, but the inner ones won't be.

Let's go with preallocation for now then, so we can make progress.

Reply:
We discussed it and agree on a path forward for now!